File: app/src/main/java/com/marvelcomics/brito/marvelcomics/ui/fragment/series/SeriesFragment.java
package com.marvelcomics.brito.marvelcomics.ui.fragment.series;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.marvelcomics.brito.entity.SeriesEntity;
import com.marvelcomics.brito.marvelcomics.R;
import com.marvelcomics.brito.marvelcomics.databinding.FragmentSeriesBinding;
import com.marvelcomics.brito.marvelcomics.ui.activity.home.HomeActivity;
import com.marvelcomics.brito.marvelcomics.ui.fragment.ItemOffSetDecorationHorizontal;
import com.marvelcomics.brito.presentation.presenter.home.HomeContract;
import com.marvelcomics.brito.presentation.presenter.series.SeriesContract;
import com.marvelcomics.brito.presentation.presenter.series.SeriesPresenter;

import java.util.List;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

public class SeriesFragment extends Fragment implements SeriesContract.View {

    private static final String ARGUMENT_CHARACTER_ID = "character_id_args";
    private static final String ARGUMENT_HOME_VIEW = "home_view_args";
    private int characterId;

    @Inject
    protected SeriesPresenter seriesPresenter;

    private FragmentSeriesBinding binding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        characterId = args.getInt(ARGUMENT_CHARACTER_ID);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        AndroidSupportInjection.inject(this);
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_series, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        loadSeries();
    }

    @Override
    public void showSeries(List<SeriesEntity> seriesEntities) {
        binding.progressbarLoadingSeries.setVisibility(View.GONE);
        binding.recyclerviewFragmentSeries.setVisibility(View.VISIBLE);
        createdAdapter(seriesEntities);
    }

    @Override
    public void showError(String message) {
        //TODO getActivity.getTheme could return nullPointer
        binding.progressbarLoadingSeries.setVisibility(View.GONE);
        binding.recyclerviewFragmentSeries.setVisibility(View.VISIBLE);
        //AlertDialogUtils.showSimpleDialog("Erro", message, getActivity());
    }

    private void loadSeries() {
        seriesPresenter.loadSeries(characterId);
    }

    private void createdAdapter(List<SeriesEntity> seriesResource) {
        SeriesAdapter adapter = new SeriesAdapter(seriesResource);
        binding.recyclerviewFragmentSeries.setLayoutManager(
                new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        binding.recyclerviewFragmentSeries.setAdapter(adapter);
        binding.recyclerviewFragmentSeries.addItemDecoration(new ItemOffSetDecorationHorizontal(8));
    }

    public static SeriesFragment newInstance(int characterId) {
        SeriesFragment fragment = new SeriesFragment();

        Bundle args = new Bundle();
        args.putInt(ARGUMENT_CHARACTER_ID, characterId);

        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/com/marvelcomics/brito/view/home/fragment/series/SeriesFragment.kt
package com.marvelcomics.brito.view.home.fragment.series

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.marvelcomics.brito.R
import com.marvelcomics.brito.databinding.FragmentSeriesBinding
import com.marvelcomics.brito.domain.entity.SeriesEntity
import com.marvelcomics.brito.domain.usecase.CoroutineUseCase
import com.marvelcomics.brito.infrastructure.utils.AlertDialogUtils
import com.marvelcomics.brito.presentation.series.SeriesInteraction
import com.marvelcomics.brito.presentation.series.SeriesScreenState
import com.marvelcomics.brito.presentation.series.SeriesViewModel
import com.marvelcomics.brito.view.extensions.viewBinding
import com.marvelcomics.brito.view.home.fragment.ItemOffSetDecorationHorizontal
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject

class SeriesFragment : Fragment(R.layout.fragment_series) {

    private val binding by viewBinding(FragmentSeriesBinding::bind)
    private val seriesViewModel: SeriesViewModel by inject()

    private var characterId: Int? = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        characterId = arguments?.getInt(ARGUMENT_CHARACTER_ID)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        loadSeries()
    }

    override fun onStart() {
        super.onStart()
        initObservers()
    }

    private fun initObservers() {
        lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                seriesViewModel.bind().collect {
                    when (it) {
                        is SeriesScreenState.Success -> {
                            showSeries(
                                CoroutineUseCase.castSuccess(it.data)
                            )
                        }
                        is SeriesScreenState.Error -> {
                            it.exception.message?.let { message ->
                                showError(message)
                            }
                        }
                        is SeriesScreenState.Loading -> {
                            showLoading()
                        }
                        is SeriesScreenState.NetworkError -> {
                            // do nothing
                        }
                    }
                }
            }
        }
    }

    private fun loadSeries() {
        seriesViewModel.handle(SeriesInteraction.LoadSeriesById(characterId ?: let { 0 }))
    }

    private fun showSeries(series: List<SeriesEntity>) {
        binding.progressbarLoadingSeries.visibility = View.GONE
        binding.recyclerviewFragmentSeries.visibility = View.VISIBLE
        createAdapter(series)
    }

    private fun showLoading() {
        binding.progressbarLoadingSeries.visibility = View.VISIBLE
        binding.recyclerviewFragmentSeries.visibility = View.GONE
    }

    private fun showError(message: String) {
        binding.progressbarLoadingSeries.visibility = View.GONE
        binding.recyclerviewFragmentSeries.visibility = View.VISIBLE
        AlertDialogUtils.showSimpleDialog("Erro", message, requireContext())
    }

    private fun createAdapter(listSeries: List<SeriesEntity>) {
        val seriesAdapter = SeriesAdapter(listSeries)
        binding.recyclerviewFragmentSeries.layoutManager =
            LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        binding.recyclerviewFragmentSeries.adapter = seriesAdapter
        binding.recyclerviewFragmentSeries.addItemDecoration(ItemOffSetDecorationHorizontal(8))
    }

    companion object {
        const val ARGUMENT_CHARACTER_ID = "character_id_args"

        fun newInstance(characterId: Int): SeriesFragment {
            val fragment = SeriesFragment()
            val args = Bundle()
            args.putInt(ARGUMENT_CHARACTER_ID, characterId)

            fragment.arguments = args
            return fragment
        }
    }
}
